### PR TITLE
Implementation of dns-01 challenge support in Confconsole Let's Encrypt plugin.

### PR DESCRIPTION
**Pull request by NitrogenUA**
Challenge type and DNS provider are new parameters introduced in dehydrated-wrapper. Use of these parameters will manipulate dehydrated configuration accordingly.
DNS manipulation is performed by [lexicon](https://github.com/AnalogJ/lexicon).
DNS provider authorization is configured by the user manually, via lexicon.yml configuration file.

### CODE REVIEW

**Comment by JedMeister** [comment]
This looks great @NitrogenUA. Bonus points for cleaning up the whitespace and fixing my blunder with the `$FULLCHAINFILE` variable name!

I think that this is a pretty cool way to go. In theory, I'm on board.

My only reservation is that it adds about 6.5MB (of dependencies) to the image. Obviously 6.5 MB on a single ISO is negligible. But the appliances have been bloating bigger and bigger every release and making them bigger always raises my anxiety levels... Having said that, it's not a ridiculous size increase and if the functionality makes user's life easier, then that's a win.

Having said that, I wonder if we could perhaps just install the additional dependencies if a user chooses to set up ACME updates via DNS?

Also FYI, there is no need to support old v16.x appliances (based on Debian 10/Buster). As you've likely noticed, we're currently working on v17.x release (based on Debian 11/Bullseye). We won't be releasing an updated Confconsole for v16.x (unless something breaks really badly), so it's totally reasonable to focus on Bullseye only.

Also, it'd be great to get some thoughts from my colleagues:

@qrntz @OnGle @Christie018 - could you all please have a look and share your thoughts.

**Comment by NitrogenUA** [comment]
@JedMeister Good point! I agree, having it optional sounds like a good compromise. Not everyone will use it, so it's not essential to carry the dependencies.

**Comment by JedMeister** [comment]
@NitrogenUA - Ok cool.

It looks like you are already handling the possibility that the deps are missing (https://github.com/turnkeylinux/confconsole/pull/64/files#diff-419a289227d16ea5e7978706c85626b5ec3ccfe122828fa4a0268d7923ed90f9R78-R79). My only suggestion there is that you note that `nslookup` is provided by the `dnsutils` package (as `lexicon` is provided by the `lexicon` package that probably doesn't need to be explicitly noted).

I suggest that the dependency packages are moved from `Depends` (https://github.com/turnkeylinux/confconsole/blob/4805133be3dc5f6ff3c03600572d2d9b07b9c164/debian/control#L14) to `Recommends` (https://github.com/turnkeylinux/confconsole/blob/4805133be3dc5f6ff3c03600572d2d9b07b9c164/debian/control#L22).

FYI we install with `--no-recommends` by default, so we'll still need to handle the case where they aren't installed (I suggest we do that in Confconsole).

**Comment by NitrogenUA** [comment]
@JedMeister  Finished rough implementation and it has performed successfully as far as my tests went, including with multiple domains per certificate.
Perhaps you could help me address some of the concerns I currently have.

- I was unable to expose dialog's `editbox()` method, insofar that it refused to open the lexicon's config file. And I'm not sure how to debug that. Therefore present implementation lacks direct lexicon config editing, which has to be performed outside of the Confconsole for the time being. I can potentially hack around that by employing `confconsole.form()`, similar to how it is used for domain names input. Though I have not yet examined how many lines should typically be enough for most providers and that number may change as lexicon inevitably expands, so hard cap would not be very future-proof.

- Would it be more optimal to implement dns-01 challenge as a dedicated menu entry within Let's Encrypt plugin? My thinking was that the dialogue prompts would be largely the same except for the next example maybe and it's still doing the same thing - obtaining a certificate. Therefore I have implemented it under existing entry.

- Relative placement of UI prompts can be improved? For instance it is not technically necessary to display 'Have you configured the DNS?' notice in dns-01 scenario. As dns-01 challenge doesn't seem to require the existence of A-records to succeed. Would it be fair to move it past the challenge type selection and display only for http-01 challenge? Or somehow split it in two as it also contains a warning regarding rate-limiting? I don't personally think it's such a huge distraction where it currently sits though.

- [Minor] For the time being I have suppressed apt stdout for dependencies installation. Should it instead be displayed to make installation progress more obvious?

File: plugins.d/Lets_Encrypt/get_certificate.py [hunk 1]
```diff
@@ -73,6 +74,10 @@ def invalid_domains(domains):
                 return ('Error in {}: Domain names must not exceed 254'
                         ' characters'.format(domain))
             for part in domain.split('.'):
+                if not len(domain.split('.')) > 1:
```
Comment (OnGle) [inline plugins.d/Lets_Encrypt/get_certificate.py:77]: This check seems a little off

- checks domain segment count, but inside an iteration over segments. If there are 0 segments (which it appears should be an error case) this check will never occur. Best to lift it out of the `for part in domain.split('.')` loop
- instead of `not segment_count > 1`, `segment_count < 2` would be better
- `domain.count('.')` would be a more succinct way to count segments than `len(domain.split('.'))`

**Review by OnGle — CHANGES_REQUESTED** [review]
Apologies for the wait on this.

Only commenting on the code, not the actual functionality behind it here. Beyond the issue noted on `get_certificate.py:77` this looks good to me.

**Comment by OnGle** [comment]
> I was unable to expose dialog's `editbox()` method, insofar that it refused to open the lexicon's config file. And I'm not sure how to debug that. Therefore present implementation lacks direct lexicon config editing, which has to be performed outside of the Confconsole for the time being. I can potentially hack around that by employing `confconsole.form()`, similar to how it is used for domain names input. Though I have not yet examined how many lines should typically be enough for most providers and that number may change as lexicon inevitably expands, so hard cap would not be very future-proof.

Hmm, I might have a look at this, leveraging `confconsole.form()` would be better than nothing but `editbox()` would be ideal. And yeah from my experience working with dialog can be a bit painful, when it works it works, when it doesn't it's a real pain to debug.

> [Minor] For the time being I have suppressed apt stdout for dependencies installation. Should it instead be displayed to make installation progress more obvious?

I think active feedback is good for UX. Long-term it'd be nice to wrap this in a `progressbox` to keep everything looking the same, but unless this functionality is already exposed this would probably be more work than it's worth, so for now I say this is good enough.

I'll leave the others for @JedMeister as he knows more about the DNS & LE specific stuff.

File: plugins.d/Lets_Encrypt/get_certificate.py (hunk 1 above)
Comment (JedMeister) [reply plugins.d/Lets_Encrypt/get_certificate.py:77]: @NitrogenUA - great idea to add additional domain validation here. However, as @OnGle notes, it should probably be before the `for` loop, instead of inside it.

@OnGle - Thanks! I wasn't aware of `count()` - looks like a very handy string method. And certainly looks like a great way to validate the number of domain segments here.

**Comment by JedMeister** [comment]
Hey @NitrogenUA - good work. Other than a minor point (re use of `apt`) and @OnGle's point re the domain validation you've added, I'm happy with your code.

>I was unable to expose dialog's editbox() method, insofar that it refused to open the lexicon's config file. And I'm not sure how to debug that. Therefore present implementation lacks direct lexicon config editing, which has to be performed outside of the Confconsole for the time being. I can potentially hack around that by employing confconsole.form(), similar to how it is used for domain names input. Though I have not yet examined how many lines should typically be enough for most providers and that number may change as lexicon inevitably expands, so hard cap would not be very future-proof.

As both yourself and @OnGle noted, `editbox()` would certainly be a better option, but `form()` will do for now.

FWIW I have started refactoring and improving our python dialog wrapper with the intention of generalising it into a common library and removing it from confconsole. Then the same code could be shared with [inithooks](https://github.com/turnkeylinux/inithooks) (which has it's own dialog-wrapper - why I don't know, it was well before my time). But as you can imagine, that's not super high priority.

>Would it be more optimal to implement dns-01 challenge as a dedicated menu entry within Let's Encrypt plugin? My thinking was that the dialogue prompts would be largely the same except for the next example maybe and it's still doing the same thing - obtaining a certificate. Therefore I have implemented it under existing entry.

Either would work, but I think your thoughts make sense. I'm happy with them being within the one menu entry.

>Relative placement of UI prompts can be improved? For instance it is not technically necessary to display 'Have you configured the DNS?' notice in dns-01 scenario. As dns-01 challenge doesn't seem to require the existence of A-records to succeed. Would it be fair to move it past the challenge type selection and display only for http-01 challenge? Or somehow split it in two as it also contains a warning regarding rate-limiting? I don't personally think it's such a huge distraction where it currently sits though.

Yeah, the DNS note isn't really relevant if using "DNS-01", so I think the user should only see that if they select "HTTP-01".

>[Minor] For the time being I have suppressed apt stdout for dependencies installation. Should it instead be displayed to make installation progress more obvious?

I'm with @OnGle on this one too. UX feedback is good, but I'm fine with it being suppressed.

**Comment by Christie018** [comment]
This definitely seems like it has been a hand full as far as the updating process goes. For the most part, the code seems finished, just needs a couple last tweaks that everyone has already pointed out:
     1. Domain validation code tweaks in the get_certificate.py file.in line 77. My college @OnGle noted that this specific code should be switched around to take this code outside of the existing for loop statement.
     2. Changing all scripts to use apt-get instead of apt to install the necessary packages. This gives off the occasional error messages when ran in script and follows the coding procedures like all of the TurnKey applications 
     3. Figure out if you want to change the dns-01 challenge menu entry within Let's Encrypt plugin.
In summary, most of the code changes in the PR have been implemented and reviewed by the team now, it's just these last couple issues in the code that need to be changed and documented in the changelog.

File: plugins.d/Lets_Encrypt/get_certificate.py [hunk 2]
```diff
@@ -132,6 +137,51 @@ def run():
...
+        elif provider == 'auto' and not which('nslookup'):
+            ret = console.yesno(
+                'nslookup tool is required to use dns-01 challenge with auto provider.\n\n'
+                'Would you like to install it now?',
+                autosize=True
+            )
+            if ret != 'ok':
+                return
+
+            apt = subprocess.run(['apt', '-y', 'install', 'dnsutils'],
```
[the remark this answers is not in the document]
Comment (NitrogenUA) [reply plugins.d/Lets_Encrypt/get_certificate.py]: Fair point, I kinda forgot about that warning. Never actually scripted anything around apt, tbh.

**Comment by NitrogenUA** [comment]
@OnGle @JedMeister @Christie018 thank you for your patience and feedback! Hope this looks better now.

File: plugins.d/Lets_Encrypt/get_certificate.py (hunk 1 above)
Comment (NitrogenUA) [reply plugins.d/Lets_Encrypt/get_certificate.py]: @OnGle thanks for pointing this out while also offering a suggestion! This is a first for me in a number of ways. Admittedly I was feeling too timid, trying to add as little as possible to the detriment of clarity.

Tbh, I couldn't locate a zero-segment scenario you mentioned, outside of us deliberately using `count()` with incorrect delimiter with. But hopefully I understood you correctly and you'll find present implementation satisfactory.

**Comment by JedMeister** [comment]
Ok great. I'm happy with this now. I'll get @OnGle & @Christie018 to share their 2c as well, but I'll probably merge as soon as one of them drops in with approval.

**Review by Christie018 — COMMENTED** [review]
Overall, it the appliance looks good to me. I noted two files that would need to be updated to follow the Turnkey Linux coding standard for import statements. Additionally, the appliance's changelog file would need to be updated to reflect the code changes.
     This not only gives you credit for the work you did, but also helps keep up to date documentation for future development. After these changes are applied and the appliance has been tested, I see no reason not to merge to master branch.

File: plugins.d/Lets_Encrypt/get_certificate.py [hunk 3]
```diff
@@ -5,7 +5,8 @@
 import subprocess
 from subprocess import PIPE
 from os import path, remove
-from shutil import copyfile
+from shutil import copyfile, which
+import dns_01
```
[the remark this answers is not in the document]
Comment (NitrogenUA) [reply plugins.d/Lets_Encrypt/get_certificate.py:9]: @Christie018 would it be fair in this instance to put `dns_01` at the very end since that is a local TurnKey module this PR introduces? Unless I am misunderstanding the standard here.
Like so:
```
import subprocess

from subprocess import PIPE
from os import path, remove
from shutil import copyfile, which 

import dns_01
```

File: plugins.d/Lets_Encrypt/dns_01.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/python3
+import sys
+import subprocess
+from subprocess import PIPE
+from os import path
+from shutil import which
+import re
```
[the remark this answers is not in the document]
Comment (NitrogenUA) [reply plugins.d/Lets_Encrypt/dns_01.py:7]: How should I go about creating a changelog given this is not an appliance, but a TKL component?

[the remark this answers is not in the document]
Comment (JedMeister) [reply plugins.d/Lets_Encrypt/dns_01.py:7]: Good point. The way that it's meant to work is create a new file in `docs/` called `RelNotes-x.y.x.txt` (where `x.y.z` is the version number). But I've been slack and haven't done that since v1.0.0 release a few years ago...

Please do do that though so there's record of all this work you've done. Also, let's make the release for your updates `v2.1.0` (the current version is `2.0.5` and IMO these changes add significant functionality, enough to warrant a minor version bump IMO.

Noe, I'll need to take care of that when I build (I'll tag the repo once it's all merged in and then the version number of the package is autogenerated from the tag).

File: plugins.d/Lets_Encrypt/get_certificate.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JedMeister) [reply plugins.d/Lets_Encrypt/get_certificate.py:9]: @NitrogenUA Yep, good work. :+1: 

And @Christie018 - nice attention to detail! Love your work! :smile:

**Comment by JedMeister** [comment]
This is looking incredibly close! TBH, I'm a bit excited about it, it's going to be pretty cool IMO.

**Review by OnGle — APPROVED** [review]
Re-review after changes. All looks good on my end. Good stuff, really appreciate it @NitrogenUA!

**Comment by JedMeister** [comment]
Thanks @OnGle - I'm feeling pretty pumped about implementing this!

I'm currently a bit tied up with other backlog stuff, but once I've cleared the immediate priorities, I hope to build this ASAP. Initially I will just push it to our `bullseye-testing` repo and post a blog post about it. But assuming we don't get any bug reports, after a few weeks, I'll upload it to our main repo.

**Comment by Christie018** [comment]
> Thanks @OnGle - I'm feeling pretty pumped about implementing this!
> 
> I'm currently a bit tied up with other backlog stuff, but once I've cleared the immediate priorities, I hope to build this ASAP. Initially I will just push it to our `bullseye-testing` repo and post a blog post about it. But assuming we don't get any bug reports, after a few weeks, I'll upload it to our main repo.

Hey boss @JedMeister , I know you're super busy. Did you get a chance to test this out yet? I wasn't able to find the bullseye-testing repo you were talking about for confconsole.

**Comment by JedMeister** [comment]
Thanks for the bump @Christie018 and thanks for your work @NitrogenUA.

I did build and test it and it worked great, so I was happy. Unfortunately though, when I tried to push the updated package to the TurnKey apt repo, I broke the repo! I'm still not 100% sure what happened, but it seems likely that I accidentally signed it with the wrong key (I'm not sure how that happened as I have a script I use to do that, so it should have auto used the relevant key - but that was the issue).

Anyway, I'm going to merge this now as I've tested it and it worked great.

**Comment by JedMeister** [comment]
I have tagged a new release (v2.0.6) and I've built the package. I have also published a GitHub "release": https://github.com/turnkeylinux/confconsole/releases/tag/v2.0.6

I do plan to push this updated release to our apt repo, but I need to sit down and work out what the issue is with keys. I'm going to need to set up a new Bookworm repo shortly, so I will sort that out when I do that. Until then, to install this newer version, it will need to be manually downloaded.
